c_src/membrane_element_sdl/cnodeserver.c

```diff
@@ -0,0 +1,225 @@
+                   ei_x_buff *in_buf, State *state) {
+  ei_x_buff out_buf;
+  ei_x_new_with_version(&out_buf);
+  int decode_idx = 0;
+  int version;
+  char fun[255];
+  int arity;
+
+  if (ei_decode_version(in_buf->buff, &decode_idx, &version)) {
+    goto handle_message_error;
+  }
+  ei_decode_tuple_header(in_buf->buff, &decode_idx, &arity);
+  if (ei_decode_atom(in_buf->buff, &decode_idx, fun)) {
+    goto handle_message_error;
+  }
+
+  int res = 1;
+  if (!strcmp(fun, "create")) {
+    long width, height;
+    if (ei_decode_long(in_buf->buff, &decode_idx, &width) ||
+        ei_decode_long(in_buf->buff, &decode_idx, &height)) {
+      goto handle_message_error;
+    }
+    res = create((int)width, (int)height, state);
+  }
+
+  if (!strcmp(fun, "display_frame")) {
+    Shmex payload;
+    if (shmex_deserialize(in_buf->buff, &decode_idx, &payload)) {
+      goto handle_message_error;
+    }
+    res = display_frame(&payload, state);
+  }
+
+  if (!strcmp(fun, "destroy")) {
+    res = destroy(state);
+  }
+
+  if (!strcmp(fun, "type")) {
+    int type, size;
+    ei_get_type(in_buf->buff, &decode_idx, &type, &size);
+    fprintf(stderr, "type: %c, size: %d", type, size);
+    res = 0;
+  }
+
+  if (ei_x_encode_tuple_header(&out_buf, 2) ||
+      ei_x_encode_atom(&out_buf, node_name) ||
+      ei_x_encode_atom(&out_buf, res ? "error" : "ok")) {
+    goto handle_message_error;
+  }
+
+  ei_send(ei_fd, &emsg.from, out_buf.buff, out_buf.index);
+
+  ei_x_free(&out_buf);
+  return 0;
+
+handle_message_error:
+  ei_x_free(&out_buf);
+  fprintf(stderr, "Message handling error\r\n");
+  return 1;
+}
+
+int receive(int ei_fd, char *node_name, State *state) {
+  ei_x_buff in_buf;
+  ei_x_new(&in_buf);
+  erlang_msg emsg;
+  int res = 0;
+  event_loop();
+  switch (ei_xreceive_msg_tmo(ei_fd, &emsg, &in_buf, 100)) {
+  case ERL_TICK:
+    // fprintf(stderr, "erl_tick\r\n");
+    break;
+  case ERL_ERROR:
+    res = erl_errno != ETIMEDOUT;
+    break;
+  default:
+    if (emsg.msgtype == ERL_REG_SEND &&
+        handle_message(ei_fd, node_name, emsg, &in_buf, state)) {
+      res = -1;
+    }
+    break;
+  }
+
+  ei_x_free(&in_buf);
+  return res;
+}
+
+int validate_args(int argc, char **argv) {
+  if (argc != 6) {
+    return 1;
+  }
+  for (int i = 1; i < argc; i++) {
+    if (strlen(argv[i]) > 255) {
+      return 1;
+    }
+  }
+  return 0;
+}
+
+int main(int argc, char **argv) {
+  if (validate_args(argc, argv)) {
+    fprintf(stderr,
+            "%s <host_name> <alive_name> <node_name> <cookie> <creation>\r\n",
+            argv[0]);
+    return 1;
+  }
+  char host_name[256];
+  strcpy(host_name, argv[1]);
+  char alive_name[256];
+  strcpy(alive_name, argv[2]);
+  char node_name[256];
+  strcpy(node_name, argv[3]);
+  char cookie[256];
+  strcpy(cookie, argv[4]);
+  short creation = (short)atoi(argv[5]);
+
+  int listen_fd;
+  int port;
+  if (listen_sock(&listen_fd, &port)) {
+    fprintf(stderr, "listen error\r\n");
```

Comment:
Do we really want to always print that? Maybe that should be printed only when some macro is defined? or flag passed?

Reply:
This applies to all `fprintf(stderr..` calls